metadetect/metadetect.py

```diff
@@ -222,82 +234,214 @@ def go(self):
                 if np.all((obs.bmask & self['nodet_flags']) != 0):
                     any_all_masked = True
 
-        if not np.any(self.mfrac < 1) or any_all_zero_weight or any_all_masked:
+        if not np.any(mfrac < 1) or any_all_zero_weight or any_all_masked:
             self._result = None
             return
 
-        try:
-            odict = self._get_all_metacal(self.mbobs)
-        except BootPSFFailure:
-            odict = None
+        # we do metacal on everything so we can get fluxes for non-shear bands later
+        mcal_res = self._get_mbobs_data(None, list(range(self.nband)))["mcal_res"]
 
-        if self.nonshear_mbobs is not None:
-            try:
-                nonshear_odict = self._get_all_metacal(self.nonshear_mbobs)
-            except BootPSFFailure:
-                nonshear_odict = None
+        if mcal_res is None:
+            self._result = None
+            return
 
-        if (
-            odict is None
-            or (
-                self.nonshear_mbobs is not None
-                and nonshear_odict is None
-            )
+        # past this point, the code should always return a dictionary with the minimal
+        # metacal types
+        # this indicates that a measurement should have been possible
+        # we may find nothing, but that is a different thing
+        all_res = {}
+        for shear_bands in self._shear_band_combs:
+            if self.color_key_func is not None and self.color_dep_mbobs is not None:
+                res = self._go_bands_with_color(shear_bands, mcal_res)
+            else:
+                res = self._go_bands(shear_bands, mcal_res)
+            if res is not None:
+                for k, v in res.items():
+                    if v is None:
+                        continue
+                    if k not in all_res:
+                        all_res[k] = []
+                    all_res[k].append(v)
+
+        for k in all_res:
+            all_res[k] = np.hstack(all_res[k])
+
+        for mcal_type in self['metacal'].get(
+            "types", ngmix.metacal.METACAL_MINIMAL_TYPES
         ):
-            self._result = None
-        else:
-            self._result = {}
-            for shear_str, mbobs in odict.items():
-                if self.nonshear_mbobs is not None and nonshear_odict is not None:
-                    nonshear_mbobs = nonshear_odict[shear_str]
-                else:
-                    nonshear_mbobs = None
-                self._result[shear_str] = self._measure(
-                    mbobs, shear_str, nonshear_mbobs=nonshear_mbobs
+            if mcal_type not in all_res:
+                all_res[mcal_type] = None
```

Comment:
```suggestion
                # this can happen if we do not detect sources for one of the
                # metacal images
                all_res[mcal_type] = None
```